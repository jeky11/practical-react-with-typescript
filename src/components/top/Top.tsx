import React, { memo } from 'react';
import styled from '@emotion/styled';

import { Legend, LegendProps } from './Legend';
import { GameName, GameNameProps } from './GameName';

export type TopComponentType = LegendProps & GameNameProps;

export const Top = memo(
  ({ children, ...legendProps }: TopComponentType): JSX.Element => (
    <Header>
      <GameName>{children}</GameName>
      <Legend {...legendProps} />
    </Header>
  )
);

// Stryker disable next-line StringLiteral
Top.displayName = 'Top';

const Header = styled.header`
  text-align: center;
  position: relative;
  display: inline-block;
`;
