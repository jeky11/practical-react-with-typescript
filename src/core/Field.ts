import { incrementNeighbours } from '@/core/CellsManipulator';

export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12; //union
export type Field = Cell[][];
export type Coords = [number, number]; //tuple

//record with string as key and cell as value
export const CellState: Record<string, Cell> = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  mark: 11,
  weakFlag: 12,
};

export const emptyFieldGenerator = (
  size: number,
  state: Cell = CellState.empty
): Field => new Array(size).fill(null).map(() => new Array(size).fill(state));

export const fieldGenerator = (size: number, probability: number): Field => {
  if (probability < 0 || probability > 1) {
    throw new Error('Probability must be between 0 and 1');
  }

  let unprocessedCells = size * size;
  let restCellsWithBombs = unprocessedCells * probability;

  const result: Field = emptyFieldGenerator(size);

  // Stryker disable next-line EqualityOperator
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      // Stryker disable next-line EqualityOperator
      if (restCellsWithBombs / unprocessedCells > Math.random()) {
        result[y][x] = CellState.bomb;
        incrementNeighbours([y, x], result);
        restCellsWithBombs--;
      }
      unprocessedCells--;
    }
  }

  return result;
};
