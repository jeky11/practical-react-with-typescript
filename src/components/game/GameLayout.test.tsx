import React from 'react';
import { render } from '@testing-library/react';

import { Grid } from '@/components/grid';
import { Top } from '@/components/top';
import { Scoreboard } from '@/components/scoreboard';

import { GameOver } from '@/components/game';

import { Field } from '@/core/Field';
import { fieldGenerator } from '@/core/__mocks__/Field';

import { GameLayout } from './GameLayout';

it('Top renders correctly', () => {
  const { asFragment } = render(
    <GameLayout
      top={
        <Top feature="Flag" firstAction="ctrl" secondAction="click">
          Minesweeper
        </Top>
      }
    >
      <Scoreboard
        time="000"
        bombs="000"
        levels={['beginner', 'intermediate', 'expert']}
        onReset={() => null}
        onChangeLevel={() => null}
      />
      <GameOver onClick={() => null} isWin={true} />
      <Grid onClick={() => null} onContextMenu={() => null}>
        {fieldGenerator(9) as Field}
      </Grid>
    </GameLayout>
  );

  expect(asFragment()).toMatchSnapshot();
});
