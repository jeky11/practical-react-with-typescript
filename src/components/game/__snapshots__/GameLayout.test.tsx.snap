// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  text-align: center;
  position: relative;
  display: inline-block;
}

.emotion-2 {
  font-size: 2em;
}

.emotion-3 {
  font-size: 1em;
  margin: 0 auto 2vw;
  line-height: 1.25em;
}

.emotion-4 {
  background: #e3e3e3;
}

.emotion-5 {
  color: #ec433c;
}

.emotion-6 {
  color: #2a48ec;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: 6px solid #e3e3e3;
  background-color: #e3e3e3;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-9 {
  display: inline-block;
  padding: 0 0.3vw;
  color: #ec433c;
  border: 0.15vw inset;
  line-height: 2vw;
  letter-spacing: 0.08em;
  background: #333;
  text-shadow: 0 0 0.1vw #ec433c;
}

.emotion-10 {
  margin: 0;
  height: 100%;
  border-radius: 0;
  border: 0.15vw solid;
  border-color: white #9e9e9e #9e9e9e white;
  background-color: #d1d1d1;
}

.emotion-11 {
  font-weight: normal;
  display: block;
  white-space: nowrap;
  min-height: 1.2em;
  padding: 0 0.2vw 0.2vw;
}

.emotion-14 {
  font-size: 1.1vw;
  height: 100%;
  cursor: pointer;
  font-weight: 700;
  border-width: 0.15vw;
  border-style: solid;
  background-color: #d1d1d1;
  border-color: white #9e9e9e #9e9e9e white;
}

.emotion-16 {
  top: 60%;
  left: 50%;
  z-index: 11;
  width: 8vw;
  height: 8vw;
  font-size: 4vw;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  line-height: 8vw;
  position: absolute;
  border-radius: 50%;
  pointer-events: auto;
  background-color: #d1d1d1;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  border: 1px solid rgba(51, 51, 51, 0.25);
}

.emotion-17 {
  display: grid;
  grid-template-columns: repeat(9, auto);
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  padding: 1vw;
}

.emotion-18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  width: 1.8vh;
  height: 1.8vh;
  color: transparent;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;
  border-color: #dddddd;
  cursor: default;
  color: rgba(0,0,0,0);
}

.emotion-18:hover {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
}

.emotion-18:hover {
  -webkit-filter: brightness(1);
  filter: brightness(1);
}

.emotion-24 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  width: 1.8vh;
  height: 1.8vh;
  color: transparent;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;
  border-color: #dddddd;
  cursor: default;
  color: #2a48ec;
}

.emotion-24:hover {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
}

.emotion-24:hover {
  -webkit-filter: brightness(1);
  filter: brightness(1);
}

.emotion-25 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  width: 1.8vh;
  height: 1.8vh;
  color: transparent;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;
  border-color: #dddddd;
  cursor: default;
  color: rgba(0,0,0,0);
  background-color: #ec433c;
}

.emotion-25:hover {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
}

.emotion-25:hover {
  -webkit-filter: brightness(1);
  filter: brightness(1);
}

.emotion-26 {
  border-radius: 50%;
  width: 1vh;
  height: 1vh;
  background-color: #333;
}

.emotion-34 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  width: 1.8vh;
  height: 1.8vh;
  color: transparent;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;
  border-color: #dddddd;
  cursor: default;
  color: #2bb13d;
}

.emotion-34:hover {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
}

.emotion-34:hover {
  -webkit-filter: brightness(1);
  filter: brightness(1);
}

.emotion-52 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  width: 1.8vh;
  height: 1.8vh;
  color: transparent;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;
  border-color: #dddddd;
  cursor: default;
  color: #ec6561;
}

.emotion-52:hover {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
}

.emotion-52:hover {
  -webkit-filter: brightness(1);
  filter: brightness(1);
}

<div
    class="emotion-0"
  >
    <header
      class="emotion-1"
    >
      <h1
        class="emotion-2"
      >
        Minesweeper
      </h1>
      <legend
        class="emotion-3"
      >
        <strong>
          Flag: 
        </strong>
        <code
          class="emotion-4"
        >
          <span
            class="emotion-5"
          >
            ctrl
          </span>
          <span
            class="emotion-6"
          >
             + click
          </span>
        </code>
      </legend>
    </header>
    <div
      class="emotion-7"
    >
      <div
        class="emotion-8"
      >
        <div
          class="emotion-9"
        >
          000
        </div>
        <div>
          <select
            class="emotion-10"
          >
            <option
              class="emotion-11"
              value="beginner"
            >
              beginner
            </option>
            <option
              class="emotion-11"
              value="intermediate"
            >
              intermediate
            </option>
            <option
              class="emotion-11"
              value="expert"
            >
              expert
            </option>
          </select>
          <button
            class="emotion-14"
          >
            ðŸ™‚
          </button>
        </div>
        <div
          class="emotion-9"
        >
          000
        </div>
      </div>
      <div
        class="emotion-16"
      >
        ðŸ˜Ž
      </div>
      <div
        class="emotion-17"
        role="grid"
        size="9"
      >
        <div
          class="emotion-18"
          data-testid="0,0"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="0,1"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="0,2"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="0,3"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="0,4"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="0,5"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-24"
          data-testid="0,6"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-25"
          data-testid="0,7"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_0,7"
          />
        </div>
        <div
          class="emotion-24"
          data-testid="0,8"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-18"
          data-testid="1,0"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="1,1"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="1,2"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="1,3"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-24"
          data-testid="1,4"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="1,5"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-34"
          data-testid="1,6"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-34"
          data-testid="1,7"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-34"
          data-testid="1,8"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-18"
          data-testid="2,0"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="2,1"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="2,2"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="2,3"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-24"
          data-testid="2,4"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-25"
          data-testid="2,5"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_2,5"
          />
        </div>
        <div
          class="emotion-24"
          data-testid="2,6"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="2,7"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-25"
          data-testid="2,8"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_2,8"
          />
        </div>
        <div
          class="emotion-18"
          data-testid="3,0"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="3,1"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-24"
          data-testid="3,2"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="3,3"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-52"
          data-testid="3,4"
          role="cell"
        >
          3
        </div>
        <div
          class="emotion-34"
          data-testid="3,5"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-34"
          data-testid="3,6"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-24"
          data-testid="3,7"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="3,8"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-18"
          data-testid="4,0"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="4,1"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-24"
          data-testid="4,2"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-25"
          data-testid="4,3"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_4,3"
          />
        </div>
        <div
          class="emotion-34"
          data-testid="4,4"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-25"
          data-testid="4,5"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_4,5"
          />
        </div>
        <div
          class="emotion-24"
          data-testid="4,6"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-18"
          data-testid="4,7"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="4,8"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-24"
          data-testid="5,0"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="5,1"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-34"
          data-testid="5,2"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-24"
          data-testid="5,3"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-34"
          data-testid="5,4"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-24"
          data-testid="5,5"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="5,6"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="5,7"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="5,8"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="6,0"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-25"
          data-testid="6,1"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_6,1"
          />
        </div>
        <div
          class="emotion-24"
          data-testid="6,2"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-18"
          data-testid="6,3"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="6,4"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="6,5"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="6,6"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-24"
          data-testid="6,7"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-25"
          data-testid="6,8"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_6,8"
          />
        </div>
        <div
          class="emotion-24"
          data-testid="7,0"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="7,1"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="7,2"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="7,3"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-24"
          data-testid="7,4"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-34"
          data-testid="7,5"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-24"
          data-testid="7,6"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-52"
          data-testid="7,7"
          role="cell"
        >
          3
        </div>
        <div
          class="emotion-34"
          data-testid="7,8"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-18"
          data-testid="8,0"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="8,1"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-18"
          data-testid="8,2"
          role="cell"
        >
          0
        </div>
        <div
          class="emotion-24"
          data-testid="8,3"
          role="cell"
        >
          1
        </div>
        <div
          class="emotion-25"
          data-testid="8,4"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_8,4"
          />
        </div>
        <div
          class="emotion-34"
          data-testid="8,5"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-25"
          data-testid="8,6"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_8,6"
          />
        </div>
        <div
          class="emotion-34"
          data-testid="8,7"
          role="cell"
        >
          2
        </div>
        <div
          class="emotion-25"
          data-testid="8,8"
          role="cell"
        >
          <div
            class="emotion-26"
            data-testid="bomb_8,8"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
